### ISSUE TITLE
consider small change to mask implied by `is_causal=True` in SDPA when query length != kv length

### ISSUE THREAD

**Issue by nairbv** [pytorch/pytorch#115111]
### 🚀 The feature, motivation and pitch

It's pretty common in causal language models to only want the last (newly generated) token from a model. E.g.:

```
In [45]: query = torch.randn( (B, S, Eh) )
In [46]: key = torch.randn( (B, S, Eh) )
In [47]: value = torch.randn( (B, S, Eh) )

In [48]: F.scaled_dot_product_attention(query, key, value, is_causal=True )[:, -1, :]
Out[48]: tensor([[ 1.1106,  0.9176,  0.8358,  0.0195, -0.2215,  0.0049,  0.3020, -0.1529]])
```

This should be equivalent to just querying for the token you want, which one might expect to be able to do with something like:
`F.scaled_dot_product_attention(query[:, -1, :], key, value, is_causal=True )`.

Unfortunately, the offset for the implied causal mask assumes that the first index of query along the sequence dimension is aligned with the first index along the k/v dimension, and so only attends to the first input token from k/v (you can actually pass anything for query and will always get the same "incorrect" result). To make this work in the way desired, instead you'd need to use something like:

```
In [50]: F.scaled_dot_product_attention(query[:,-1,:], key, value, attn_mask=torch.ones( (1, S) ) )
Out[50]:
tensor([[[ 1.1106,  0.9176,  0.8358,  0.0195, -0.2215,  0.0049,  0.3020,
          -0.1529]]])
```
If the mask implied by is_causal=True was such that the last index of the query was assumed to be aligned with the last index of the key/value, I think that'd be more frequently useful than what's currently implied by `is_causal=True` (and I believe would also allow for things like varied layer sizes across LLMs without needing to manually construct masks).

Changing the mask implied by is_causal would be BC-breaking, but sdpa is documented as beta, and I doubt there's much code relying on this specific behavior.

In the example implementation in the documentation, I believe this would simply be equivalent to changing `temp_mask = torch.ones(L, S, dtype=torch.bool).tril(diagonal=0)` to `temp_mask = torch.ones(L, S, dtype=torch.bool).tril(diagonal=S-L)`.

When S=L, the behavior wouldn't change.

### Alternatives

pass an explicit mask

**Comment by drisspg** [comment]
Have been working on this for a little bit, hoping to land this soon:
https://github.com/pytorch/pytorch/pull/114823

**Comment by drisspg** [comment]
See: https://github.com/pytorch/pytorch/issues/108108

**Comment by nairbv** [comment]
ah cool! Thanks, closing this as duplicate.
